osu.Game/Skinning/DefaultSkin.cs

```diff
@@ -23,17 +27,77 @@ public DefaultSkin(IStorageResourceProvider resources)
 
         [UsedImplicitly(ImplicitUseKindFlags.InstantiatedWithFixedConstructorSignature)]
         public DefaultSkin(SkinInfo skin, IStorageResourceProvider resources)
-            : base(skin)
+            : base(skin, resources)
         {
             Configuration = new DefaultSkinConfiguration();
         }
 
-        public override Drawable GetDrawableComponent(ISkinComponent component) => null;
-
         public override Texture GetTexture(string componentName, WrapMode wrapModeS, WrapMode wrapModeT) => null;
 
         public override ISample GetSample(ISampleInfo sampleInfo) => null;
 
+        public override Drawable GetDrawableComponent(ISkinComponent component)
+        {
+            if (base.GetDrawableComponent(component) is Drawable c)
+                return c;
+
+            switch (component)
+            {
+                case SkinnableTargetComponent target:
+                    switch (target.Target)
+                    {
+                        case SkinnableTarget.MainHUDComponents:
+                            var skinnableTargetWrapper = new SkinnableTargetWrapper(container =>
```

Comment:
Importantly, this is never actually reached if the user has already customised the skin (even for the `DefaultSkin`). The user's layout will be returned in the `base.` call above.

Reply:
Yeah I was going to touch on this, I'm not sure if you intentionally avoided it, but since the order of the drawable component retrieval has been flipped around to prefer customized components over default, it all feels like the skinnable target content logic isn't really suitable for this.

Would like to hear your opinion in moving all this into new separate methods specific for it (and obsolete/remove `GetDrawableComponent` if that's the way forward after the transition completes, which I presume to happen):
```csharp
class Skin
{
    // returns customized skin components for given target, 
    // or GetDefaultSkinComponents(target) if there isn't.
    public SkinnableTargetWrapper GetSkinComponents(SkinnableTarget target);
    
    // returns default skin components for given target.
    protected abstract SkinnableTargetWrapper GetDefaultSkinComponents(SkinnableTarget target);
}
```

(not sure about using `SkinnableTargetWrapper` as return type outright though, maybe `Container` instead?)

Here's a branch https://github.com/ppy/osu/compare/master...frenzibyte:refactor-skin-components-retrieval if it makes it any easier to do it.

Reply:
I'm not sure what you mean by "doesn't feel suited"? Is there some expectation that the base lookup should be last? What's the rationale here?

Reply:
`GetDrawableComponent()` is usually known to call its base method last when no matching component is found, so it looks unusual to suddenly make it called first, though I wouldn't really call this a valid reason. 

But, I think defining specific methods for this would be better to avoid any confusion/misunderstanding on the `SkinnableTargetComponent` case handling in skin implementations, for the fact they return the default components and will be overridden by customized components in base method.

Either way, at this point, this may need more opinions, as I'm not really sure which way is better anymore.

Reply:
If the change you propose is to be made, it should be on the existing method, not a new one. There's no reason to split it out into its own method unless we decide to do the same for the other lookup types, or decide we don't want the type-based lookup in the first place.